File: sdk-go/dephealth/scheduler.go
package dephealth

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"sync"
	"time"
)

// Sentinel errors for the scheduler.
var (
	ErrAlreadyStarted = errors.New("scheduler already started")
	ErrNotStarted     = errors.New("scheduler not started")
)

// endpointState holds the health check state for a specific endpoint.
type endpointState struct {
	mu                   sync.Mutex
	healthy              *bool // nil = UNKNOWN (before first check)
	consecutiveFailures  int
	consecutiveSuccesses int

	// Fields for HealthDetails() API.
	lastStatus    StatusCategory
	lastDetail    string
	lastLatency   time.Duration
	lastCheckedAt time.Time

	// Static fields set at state creation time.
	depName  string
	depType  DependencyType
	host     string
	port     string
	critical bool
	labels   map[string]string
}

// Scheduler manages periodic execution of health checks.
type Scheduler struct {
	deps    []scheduledDep
	metrics *MetricsExporter
	logger  *slog.Logger

	states  map[string]*endpointState // key: "name:host:port"
	cancel  context.CancelFunc
	wg      sync.WaitGroup
	started bool
	stopped bool
	mu      sync.Mutex
}

// scheduledDep contains a dependency with its associated checker.
type scheduledDep struct {
	dep     Dependency
	checker HealthChecker
}

// SchedulerOption is a functional option for Scheduler.
type SchedulerOption func(*schedulerConfig)

type schedulerConfig struct {
	logger *slog.Logger
}

// WithSchedulerLogger sets the logger for the scheduler.
func WithSchedulerLogger(l *slog.Logger) SchedulerOption {
	return func(c *schedulerConfig) {
		c.logger = l
	}
}

// NewScheduler creates a new scheduler.
// metrics is the metrics exporter used for recording health check results.
func NewScheduler(metrics *MetricsExporter, opts ...SchedulerOption) *Scheduler {
	cfg := schedulerConfig{
		logger: slog.Default(),
	}
	for _, o := range opts {
		o(&cfg)
	}

	return &Scheduler{
		metrics: metrics,
		logger:  cfg.logger,
	}
}

// Start launches periodic health checks for all registered dependencies.
// Each endpoint of each dependency is checked in a separate goroutine.
// Calling Start more than once returns an error.
func (s *Scheduler) Start(ctx context.Context) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	if s.started {
		return ErrAlreadyStarted
	}
	s.started = true

	ctx, s.cancel = context.WithCancel(ctx)

	s.states = make(map[string]*endpointState)
	for _, sd := range s.deps {
		critical := sd.dep.Critical != nil && *sd.dep.Critical
		for _, ep := range sd.dep.Endpoints {
			key := sd.dep.Name + ":" + ep.Host + ":" + ep.Port
			labels := make(map[string]string, len(ep.Labels))
			for k, v := range ep.Labels {
				labels[k] = v
			}
			st := &endpointState{
				lastStatus: StatusUnknown,
				lastDetail: "unknown",
				depName:    sd.dep.Name,
				depType:    sd.dep.Type,
				host:       ep.Host,
				port:       ep.Port,
				critical:   critical,
				labels:     labels,
			}
			s.states[key] = st
			s.wg.Add(1)
			go s.runEndpointLoop(ctx, sd.dep, ep, sd.checker, st)
		}
	}

	return nil
}

// Stop stops the scheduler and waits for all goroutines to finish.
// Repeated calls are no-op.
func (s *Scheduler) Stop() {
	s.mu.Lock()
	if s.stopped || !s.started {
		s.mu.Unlock()
		return
	}
	s.stopped = true
	cancel := s.cancel
	s.mu.Unlock()

	cancel()
	s.wg.Wait()
}

// Health returns the current health state of all endpoints.
// Key is "dependency:host:port", value is true (healthy) / false (unhealthy).
// Endpoints in UNKNOWN state (before first check) are not included in the result.
func (s *Scheduler) Health() map[string]bool {
	s.mu.Lock()
	states := s.states
	s.mu.Unlock()

	if states == nil {
		return nil
	}

	result := make(map[string]bool, len(states))
	for key, st := range states {
		st.mu.Lock()
		if st.healthy != nil {
			result[key] = *st.healthy
		}
		st.mu.Unlock()
	}
	return result
}

// HealthDetails returns the detailed health state of all endpoints.
// Key is "dependency:host:port". Unlike Health(), UNKNOWN endpoints
// (before first check) are included with Status="unknown" and Healthy=nil.
// Returns nil before Start() is called.
func (s *Scheduler) HealthDetails() map[string]EndpointStatus {
	s.mu.Lock()
	states := s.states
	s.mu.Unlock()

	if states == nil {
		return nil
	}

	result := make(map[string]EndpointStatus, len(states))
	for key, st := range states {
		st.mu.Lock()
		es := EndpointStatus{
			Healthy:       copyBoolPtr(st.healthy),
			Status:        st.lastStatus,
			Detail:        st.lastDetail,
			Latency:       st.lastLatency,
			Type:          st.depType,
			Name:          st.depName,
			Host:          st.host,
			Port:          st.port,
			Critical:      st.critical,
			LastCheckedAt: st.lastCheckedAt,
			Labels:        copyStringMap(st.labels),
		}
		st.mu.Unlock()
		result[key] = es
	}
	return result
}

// copyBoolPtr returns a copy of a *bool pointer.
func copyBoolPtr(b *bool) *bool {
	if b == nil {
		return nil
	}
	v := *b
	return &v
}

// copyStringMap returns a shallow copy of a string map.
func copyStringMap(m map[string]string) map[string]string {
	if m == nil {
		return map[string]string{}
	}
	cp := make(map[string]string, len(m))
	for k, v := range m {
		cp[k] = v
	}
	return cp
}

// runEndpointLoop is the main check loop for a single endpoint.
func (s *Scheduler) runEndpointLoop(ctx context.Context, dep Dependency, ep Endpoint, checker HealthChecker, state *endpointState) {
	defer s.wg.Done()

	logAttrs := []slog.Attr{
		slog.String("dependency", dep.Name),
		slog.String("type", string(dep.Type)),
		slog.String("host", ep.Host),
		slog.String("port", ep.Port),
	}

	// initialDelay.
	if dep.Config.InitialDelay > 0 {
		select {
		case <-ctx.Done():
			return
		case <-time.After(dep.Config.InitialDelay):
		}
	}

	// First check.
	s.executeCheck(ctx, dep, ep, checker, state, logAttrs, true)

	// Periodic checks.
	ticker := time.NewTicker(dep.Config.Interval)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			s.executeCheck(ctx, dep, ep, checker, state, logAttrs, false)
		}
	}
}

// executeCheck performs a single health check and updates state and metrics.
func (s *Scheduler) executeCheck(
	ctx context.Context,
	dep Dependency,
	ep Endpoint,
	checker HealthChecker,
	state *endpointState,
	logAttrs []slog.Attr,
	isFirst bool,
) {
	// Skip check if context is already cancelled.
	if ctx.Err() != nil {
		return
	}

	// Create a context with timeout for the check.
	checkCtx, checkCancel := context.WithTimeout(ctx, dep.Config.Timeout)
	defer checkCancel()

	start := time.Now()
	checkErr := s.safeCheck(checkCtx, checker, ep)
	duration := time.Since(start)

	// Record latency always (both on success and failure).
	s.metrics.ObserveLatency(dep, ep, duration)

	// Classify the check result for status metrics.
	result := classifyError(checkErr)
	s.metrics.SetStatus(dep, ep, result.Category)
	s.metrics.SetStatusDetail(dep, ep, result.Detail)

	state.mu.Lock()
	defer state.mu.Unlock()

	// Store classification results for HealthDetails() API.
	state.lastStatus = result.Category
	state.lastDetail = result.Detail
	state.lastLatency = duration
	state.lastCheckedAt = time.Now()

	if isFirst {
		// First check: set state immediately without threshold logic.
		healthy := checkErr == nil
		state.healthy = &healthy
		if healthy {
			state.consecutiveSuccesses = 1
			state.consecutiveFailures = 0
		} else {
			state.consecutiveFailures = 1
			state.consecutiveSuccesses = 0
		}
		if healthy {
			s.metrics.SetHealth(dep, ep, 1)
		} else {
			s.metrics.SetHealth(dep, ep, 0)
			s.logger.LogAttrs(ctx, slog.LevelWarn, "dephealth: check failed",
				append(logAttrs, slog.String("error", checkErr.Error()))...)
		}
		return
	}

	if checkErr != nil {
		// Failed check.
		state.consecutiveSuccesses = 0
		state.consecutiveFailures++

		s.logger.LogAttrs(ctx, slog.LevelWarn, "dephealth: check failed",
			append(logAttrs, slog.String("error", checkErr.Error()))...)

		if state.healthy != nil && *state.healthy &&
			state.consecutiveFailures >= dep.Config.FailureThreshold {
			// Transition HEALTHY -> UNHEALTHY.
			healthy := false
			state.healthy = &healthy
			s.metrics.SetHealth(dep, ep, 0)
			s.logger.LogAttrs(ctx, slog.LevelError, "dephealth: dependency unhealthy",
				append(logAttrs, slog.Int("consecutive_failures", state.consecutiveFailures))...)
		} else if state.healthy != nil && !*state.healthy {
			// Already unhealthy — update metric (it is already 0).
			s.metrics.SetHealth(dep, ep, 0)
		}
	} else {
		// Successful check.
		state.consecutiveFailures = 0
		state.consecutiveSuccesses++

		if state.healthy != nil && !*state.healthy &&
			state.consecutiveSuccesses >= dep.Config.SuccessThreshold {
			// Transition UNHEALTHY -> HEALTHY.
			healthy := true
			state.healthy = &healthy
			s.metrics.SetHealth(dep, ep, 1)
			s.logger.LogAttrs(ctx, slog.LevelInfo, "dephealth: dependency recovered", logAttrs...)
		} else if state.healthy != nil && *state.healthy {
			// Already healthy — update metric (it is already 1).
			s.metrics.SetHealth(dep, ep, 1)
		}
	}
}

// safeCheck calls checker.Check with panic recovery.
func (s *Scheduler) safeCheck(ctx context.Context, checker HealthChecker, ep Endpoint) (err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("panic in health checker: %v", r)
			s.logger.Error("dephealth: panic in health checker",
				"endpoint", ep.Host+":"+ep.Port,
				"panic", r,
			)
		}
	}()
	return checker.Check(ctx, ep)
}

File: sdk-python/dephealth/scheduler.py
"""Check scheduler: periodic dependency health checks."""

from __future__ import annotations

import asyncio
import logging
import threading
import time
from dataclasses import dataclass, field
from datetime import UTC, datetime

from dephealth.check_result import STATUS_UNKNOWN, classify_error
from dephealth.checker import CheckError, HealthChecker
from dephealth.dependency import CheckConfig, Dependency, DependencyType, Endpoint
from dephealth.endpoint_status import EndpointStatus
from dephealth.metrics import MetricsExporter

logger = logging.getLogger("dephealth.scheduler")


class EndpointNotFoundError(Exception):
    """Raised when a dynamic endpoint operation targets a non-existent endpoint."""

    def __init__(self, dep_name: str, host: str, port: str) -> None:
        self.dep_name = dep_name
        self.host = host
        self.port = port
        super().__init__(f"Endpoint not found: {dep_name}:{host}:{port}")


@dataclass
class _EndpointState:
    """Health check state of a single endpoint."""

    healthy: bool | None = None
    consecutive_failures: int = 0
    consecutive_successes: int = 0

    # Dynamic fields updated after each check (for HealthDetails API).
    last_status: str = STATUS_UNKNOWN
    last_detail: str = "unknown"
    last_latency: float = 0.0
    last_checked_at: datetime | None = None

    # Static fields set at creation time.
    dep_name: str = ""
    dep_type: str = ""
    host: str = ""
    port: str = ""
    critical: bool = False
    labels: dict[str, str] = field(default_factory=dict)


@dataclass
class _SchedulerEntry:
    """Scheduler entry: dependency + checker + endpoint states."""

    dep: Dependency
    checker: HealthChecker
    states: dict[str, _EndpointState] = field(default_factory=dict)


class CheckScheduler:
    """Dependency health check scheduler.

    Supports two modes:
    - asyncio (primary): via asyncio.create_task
    - threading (fallback): via threading.Timer
    """

    def __init__(
        self,
        metrics: MetricsExporter,
        global_config: CheckConfig | None = None,
        log: logging.Logger | None = None,
    ) -> None:
        self._metrics = metrics
        self._global_config = global_config or CheckConfig(initial_delay=0)
        self._log = log or logger

        self._entries: list[_SchedulerEntry] = []
        self._lock = threading.Lock()

        # Per-endpoint state, keyed by "dep_name:host:port".
        self._states: dict[str, _EndpointState] = {}

        # Per-endpoint tracking for asyncio mode.
        self._ep_tasks: dict[str, asyncio.Task[None]] = {}

        # Per-endpoint tracking for threading mode.
        self._ep_threads: dict[str, threading.Thread] = {}
        self._ep_stop_events: dict[str, threading.Event] = {}

        self._started = False
        self._stopped = False
        self._mode: str = ""  # "async" or "sync"

    def add(self, dep: Dependency, checker: HealthChecker) -> None:
        """Add a dependency for monitoring (called before start)."""
        entry = _SchedulerEntry(dep=dep, checker=checker)
        for ep in dep.endpoints:
            hp_key = f"{ep.host}:{ep.port}"
            state = _EndpointState(
                dep_name=dep.name,
                dep_type=str(dep.type),
                host=ep.host,
                port=ep.port,
                critical=dep.critical,
                labels=dict(ep.labels),
            )
            entry.states[hp_key] = state
            state_key = f"{dep.name}:{ep.host}:{ep.port}"
            self._states[state_key] = state
        self._entries.append(entry)

    async def start(self) -> None:
        """Start in asyncio mode."""
        self._started = True
        self._stopped = False
        self._mode = "async"
        for entry in self._entries:
            for ep in entry.dep.endpoints:
                state_key = f"{entry.dep.name}:{ep.host}:{ep.port}"
                task = asyncio.create_task(self._run_loop(entry, ep))
                self._ep_tasks[state_key] = task
        self._log.info("Scheduler started (%d entries)", len(self._entries))

    async def stop(self) -> None:
        """Stop asyncio mode."""
        self._stopped = True
        with self._lock:
            tasks = list(self._ep_tasks.values())
        for task in tasks:
            task.cancel()
        await asyncio.gather(*tasks, return_exceptions=True)
        with self._lock:
            self._ep_tasks.clear()
        self._log.info("Scheduler stopped")

    def start_sync(self) -> None:
        """Start in threading mode (fallback)."""
        self._started = True
        self._stopped = False
        self._mode = "sync"
        for entry in self._entries:
            for ep in entry.dep.endpoints:
                state_key = f"{entry.dep.name}:{ep.host}:{ep.port}"
                stop_event = threading.Event()
                self._ep_stop_events[state_key] = stop_event
                t = threading.Thread(
                    target=self._run_thread,
                    args=(entry, ep, stop_event),
                    daemon=True,
                )
                self._ep_threads[state_key] = t
                t.start()
        self._log.info("Scheduler started in sync mode (%d entries)", len(self._entries))

    def stop_sync(self) -> None:
        """Stop threading mode."""
        self._stopped = True
        with self._lock:
            events = list(self._ep_stop_events.values())
            threads = list(self._ep_threads.values())
        for event in events:
            event.set()
        for t in threads:
            t.join(timeout=5.0)
        with self._lock:
            self._ep_stop_events.clear()
            self._ep_threads.clear()
        self._log.info("Scheduler stopped (sync mode)")

    def health(self) -> dict[str, bool]:
        """Return current health status of all dependencies."""
        with self._lock:
            states_snapshot = list(self._states.items())

        # Group by dep_name — a dependency is healthy if at least one endpoint is healthy.
        by_dep: dict[str, list[bool | None]] = {}
        for _key, state in states_snapshot:
            by_dep.setdefault(state.dep_name, []).append(state.healthy)

        return {name: any(h for h in healths) for name, healths in by_dep.items()}

    def health_details(self) -> dict[str, EndpointStatus]:
        """Return detailed health status of all endpoints.

        Key format: ``"dependency:host:port"``.
        Includes UNKNOWN endpoints (before first check) with status="unknown"
        and healthy=None.
        """
        with self._lock:
            states_snapshot = list(self._states.items())

        result: dict[str, EndpointStatus] = {}
        for key, state in states_snapshot:
            result[key] = EndpointStatus(
                healthy=state.healthy,
                status=state.last_status,
                detail=state.last_detail,
                latency=state.last_latency,
                type=state.dep_type,
                name=state.dep_name,
                host=state.host,
                port=state.port,
                critical=state.critical,
                last_checked_at=state.last_checked_at,
                labels=dict(state.labels),
            )
        return result

    # --- Dynamic endpoint management ---

    async def add_endpoint(
        self,
        dep_name: str,
        dep_type: DependencyType,
        critical: bool,
        ep: Endpoint,
        checker: HealthChecker,
    ) -> None:
        """Add a new endpoint at runtime (asyncio mode)."""
        self._ensure_running()
        state_key = f"{dep_name}:{ep.host}:{ep.port}"

        with self._lock:
            if state_key in self._states:
                return  # Idempotent: already exists.

            dep = Dependency(
                name=dep_name,
                type=dep_type,
                critical=critical,
                endpoints=[ep],
                config=self._global_config,
            )
            entry = _SchedulerEntry(dep=dep, checker=checker)

            state = _EndpointState(
                dep_name=dep_name,
                dep_type=str(dep_type),
                host=ep.host,
                port=ep.port,
                critical=critical,
                labels=dict(ep.labels),
            )
            hp_key = f"{ep.host}:{ep.port}"
            entry.states[hp_key] = state
            self._states[state_key] = state
            self._entries.append(entry)

            if self._mode == "async":
                task = asyncio.create_task(self._run_loop(entry, ep))
                self._ep_tasks[state_key] = task
            elif self._mode == "sync":
                stop_event = threading.Event()
                self._ep_stop_events[state_key] = stop_event
                t = threading.Thread(
                    target=self._run_thread,
                    args=(entry, ep, stop_event),
                    daemon=True,
                )
                self._ep_threads[state_key] = t
                t.start()

        self._log.info("Endpoint added: %s", state_key)

    async def remove_endpoint(self, dep_name: str, host: str, port: str) -> None:
        """Remove an endpoint at runtime (asyncio mode)."""
        self._ensure_running()
        state_key = f"{dep_name}:{host}:{port}"
        task: asyncio.Task[None] | None = None

        with self._lock:
            state = self._states.get(state_key)
            if state is None:
                return  # Idempotent: already gone.

            # Find and remove the entry.
            entry = self._find_entry(state_key)
            dep = entry.dep if entry else None
            ep = Endpoint(host=host, port=port, labels=dict(state.labels))

            if self._mode == "async":
                task = self._ep_tasks.pop(state_key, None)
                if task is not None:
                    task.cancel()
            elif self._mode == "sync":
                stop_event = self._ep_stop_events.pop(state_key, None)
                if stop_event is not None:
                    stop_event.set()
                self._ep_threads.pop(state_key, None)

            del self._states[state_key]
            if entry is not None:
                self._entries.remove(entry)

        # Await cancelled task outside lock to avoid deadlock.
        if task is not None:
            await asyncio.gather(task, return_exceptions=True)

        # Clean up metrics.
        if dep is not None:
            self._metrics.delete_metrics(dep, ep)

        self._log.info("Endpoint removed: %s", state_key)

    async def update_endpoint(
        self,
        dep_name: str,
        old_host: str,
        old_port: str,
        new_ep: Endpoint,
        checker: HealthChecker,
    ) -> None:
        """Replace an endpoint at runtime (asyncio mode)."""
        old_key = f"{dep_name}:{old_host}:{old_port}"
        with self._lock:
            if old_key not in self._states:
                raise EndpointNotFoundError(dep_name, old_host, old_port)
            old_state = self._states[old_key]
            dep_type = DependencyType(old_state.dep_type)
            critical = old_state.critical

        await self.remove_endpoint(dep_name, old_host, old_port)
        await self.add_endpoint(dep_name, dep_type, critical, new_ep, checker)

    def add_endpoint_sync(
        self,
        dep_name: str,
        dep_type: DependencyType,
        critical: bool,
        ep: Endpoint,
        checker: HealthChecker,
    ) -> None:
        """Add a new endpoint at runtime (threading mode)."""
        self._ensure_running()
        state_key = f"{dep_name}:{ep.host}:{ep.port}"

        with self._lock:
            if state_key in self._states:
                return

            dep = Dependency(
                name=dep_name,
                type=dep_type,
                critical=critical,
                endpoints=[ep],
                config=self._global_config,
            )
            entry = _SchedulerEntry(dep=dep, checker=checker)

            state = _EndpointState(
                dep_name=dep_name,
                dep_type=str(dep_type),
                host=ep.host,
                port=ep.port,
                critical=critical,
                labels=dict(ep.labels),
            )
            hp_key = f"{ep.host}:{ep.port}"
            entry.states[hp_key] = state
            self._states[state_key] = state
            self._entries.append(entry)

            stop_event = threading.Event()
            self._ep_stop_events[state_key] = stop_event
            t = threading.Thread(
                target=self._run_thread,
                args=(entry, ep, stop_event),
                daemon=True,
            )
            self._ep_threads[state_key] = t
            t.start()

        self._log.info("Endpoint added (sync): %s", state_key)

    def remove_endpoint_sync(self, dep_name: str, host: str, port: str) -> None:
        """Remove an endpoint at runtime (threading mode)."""
        self._ensure_running()
        state_key = f"{dep_name}:{host}:{port}"

        with self._lock:
            state = self._states.get(state_key)
            if state is None:
                return

            entry = self._find_entry(state_key)
            dep = entry.dep if entry else None
            ep = Endpoint(host=host, port=port, labels=dict(state.labels))

            stop_event = self._ep_stop_events.pop(state_key, None)
            if stop_event is not None:
                stop_event.set()
            thread = self._ep_threads.pop(state_key, None)

            del self._states[state_key]
            if entry is not None:
                self._entries.remove(entry)

        if thread is not None:
            thread.join(timeout=5.0)

        if dep is not None:
            self._metrics.delete_metrics(dep, ep)

        self._log.info("Endpoint removed (sync): %s", state_key)

    def update_endpoint_sync(
        self,
        dep_name: str,
        old_host: str,
        old_port: str,
        new_ep: Endpoint,
        checker: HealthChecker,
    ) -> None:
        """Replace an endpoint at runtime (threading mode)."""
        old_key = f"{dep_name}:{old_host}:{old_port}"
        with self._lock:
            if old_key not in self._states:
                raise EndpointNotFoundError(dep_name, old_host, old_port)
            old_state = self._states[old_key]
            dep_type = DependencyType(old_state.dep_type)
            critical = old_state.critical

        self.remove_endpoint_sync(dep_name, old_host, old_port)
        self.add_endpoint_sync(dep_name, dep_type, critical, new_ep, checker)

    def _ensure_running(self) -> None:
        """Raise if the scheduler is not in a running state."""
        if not self._started:
            msg = "Scheduler not started"
            raise RuntimeError(msg)
        if self._stopped:
            msg = "Scheduler already stopped"
            raise RuntimeError(msg)

    def _find_entry(self, state_key: str) -> _SchedulerEntry | None:
        """Find the _SchedulerEntry owning the given state key (lock must be held)."""
        parts = state_key.split(":", 2)
        dep_name = parts[0]
        hp_key = f"{parts[1]}:{parts[2]}"
        for entry in self._entries:
            if entry.dep.name == dep_name and hp_key in entry.states:
                return entry
        return None

    async def _run_loop(self, entry: _SchedulerEntry, ep: Endpoint) -> None:
        """Check loop for a single endpoint (asyncio)."""
        if entry.dep.config.initial_delay > 0:
            await asyncio.sleep(entry.dep.config.initial_delay)

        while not self._stopped:
            await self._check_once(entry, ep)
            await asyncio.sleep(entry.dep.config.interval)

    def _run_thread(
        self, entry: _SchedulerEntry, ep: Endpoint, stop_event: threading.Event
    ) -> None:
        """Check loop for a single endpoint (threading)."""
        if entry.dep.config.initial_delay > 0:
            stop_event.wait(entry.dep.config.initial_delay)
            if stop_event.is_set():
                return

        loop = asyncio.new_event_loop()
        try:
            while not self._stopped and not stop_event.is_set():
                loop.run_until_complete(self._check_once(entry, ep))
                stop_event.wait(entry.dep.config.interval)
        finally:
            loop.close()

    async def _check_once(self, entry: _SchedulerEntry, ep: Endpoint) -> None:
        """Run a single endpoint check and update metrics and thresholds."""
        key = f"{ep.host}:{ep.port}"
        state = entry.states[key]

        start = time.monotonic()
        check_err: BaseException | None = None
        try:
            await asyncio.wait_for(
                entry.checker.check(ep),
                timeout=entry.dep.config.timeout,
            )
            success = True
        except (CheckError, TimeoutError, Exception) as e:
            success = False
            check_err = e
            self._log.debug("Check failed for %s (%s): %s", entry.dep.name, key, e)

        duration = time.monotonic() - start
        self._metrics.observe_latency(entry.dep, ep, duration)

        # Classify the error and set status/detail metrics.
        result = classify_error(check_err)
        self._metrics.set_status(entry.dep, ep, result.category)
        self._metrics.set_status_detail(entry.dep, ep, result.detail)

        # Store classification results for health_details() API.
        state.last_status = result.category
        state.last_detail = result.detail
        state.last_latency = duration
        state.last_checked_at = datetime.now(UTC)

        if success:
            state.consecutive_successes += 1
            state.consecutive_failures = 0
            if state.consecutive_successes >= entry.dep.config.success_threshold:
                if not state.healthy:
                    self._log.info("Dependency %s (%s) is now healthy", entry.dep.name, key)
                state.healthy = True
                self._metrics.set_health(entry.dep, ep, 1.0)
        else:
            state.consecutive_failures += 1
            state.consecutive_successes = 0
            if state.consecutive_failures >= entry.dep.config.failure_threshold:
                if state.healthy:
                    self._log.warning("Dependency %s (%s) is now unhealthy", entry.dep.name, key)
                state.healthy = False
                self._metrics.set_health(entry.dep, ep, 0.0)
